### PR TITLE
Custom undo/redo implementation (#97)

### LINKED ISSUE #97: Custom undo/redo implementation
Currently code folding causes automatic creation of undo/redo records. This is done by `_TextEditingHistory` widget inside `TextField` widget.
Undoing code folding breaks the hidden state we maintain.

### PR DESCRIPTION
**Pull request by alexeyinkin**
- Resolves #97

### CODE REVIEW

File: lib/src/history/code_history_controller.dart [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+      save = code.lines.lines.length != lastCode.lines.lines.length;
+    }
+
+    if (!save) {
+      if (lastCode.text != code.text) {
+        _isTextChanged = true;
+      }
+
+      final isTextOneCharLonger = code.text.length == lastCode.text.length + 1;
+      final isSelectionChangeImportant = !isTextOneCharLonger ||
```
Comment (Malarg) [inline lib/src/history/code_history_controller.dart:56]: I think `important` has too wide sense. Also It calculates not just selection changes, but text too.

File: lib/src/code_field/text_editing_value.dart [hunk 2]
```diff
@@ -94,6 +94,14 @@ extension TextEditingValueExtension on TextEditingValue {
     );
   }
 
+  TextEditingValue typed(String text) {
```
Comment (Malarg) [inline lib/src/code_field/text_editing_value.dart:97]: This method used only in tests. May be create private extension on `TextEditingValue` there?

File: test/src/history/code_history_controller_test.dart [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+
+        expect(controller.fullText, MethodSnippet.full);
+
+        await wt.sendRedo();
+        await wt.sendRedo();
+
+        expect(controller.fullText, MethodSnippet.full + 'ab');
+      });
+
+      testWidgets('Limit depth', (WidgetTester wt) async {
```
Comment (Malarg) [inline test/src/history/code_history_controller_test.dart:297]: It will be better to split on three methods: `limitFillsAsExpected`, `undoLimitAsExpected`, `redoAsExpected`

**Review by Malarg — CHANGES_REQUESTED** [review]

File: lib/src/history/code_history_controller.dart [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+class CodeHistoryController {
+  final CodeController codeController;
+  Code lastCode;
+  TextSelection lastSelection;
+  int _currentRecordIndex = 0;
+  bool _isTextChanged = false;
+  Timer? _debounceTimer;
+
+  @visibleForTesting
+  final stack = <CodeHistoryRecord>[];
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply lib/src/history/code_history_controller.dart:30]: There is no problem calling a variable after our usage intention and not after a class. If we want to push and pop with a list, we can call it 'stack'.

File: lib/src/history/code_history_controller.dart [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  CodeHistoryController({
+    required this.codeController,
+  })  : lastCode = codeController.code,
+        lastSelection = codeController.value.selection {
+    _push();
+  }
+
+  void beforeChanged(Code code, TextSelection selection) {
+    _dropRedoIfNeed();
+    bool save = false;
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply lib/src/history/code_history_controller.dart:44]: The appropriate mode is 'should' which would turn a 4-char variable to a 10-char variable, so I would rather not.
Naming is less strict with local variables than it is with wider-scoped ones.

File: test/src/history/code_history_controller_test.dart (hunk 3 above)
Comment (alexeyinkin) [reply test/src/history/code_history_controller_test.dart:297]: Outside of tests this would sure need to be splitted.

But with tests it is more important to see and run each one as a whole. So splitting is less aggressive in tests. This is because grasping `expect` outside of `test`/`testWidgets` makes it hard to follow.

So instead of extracting the sections I added comments between them.

File: lib/src/history/code_history_controller.dart (hunk 1 above)
Comment (alexeyinkin) [reply lib/src/history/code_history_controller.dart]: Refactored.

File: lib/src/code_field/text_editing_value.dart (hunk 2 above)
Comment (alexeyinkin) [reply lib/src/code_field/text_editing_value.dart]: `+`

File: lib/src/history/code_history_controller.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (Malarg) [reply lib/src/history/code_history_controller.dart:44]: > Naming is less strict with local variables than it is with wider-scoped ones.
Considering broken windows theory it isn't. I believe, that code have to be clean, regardless it placement. Specifically at this place renaming won't create any additional line breaks, but code will be more readable

File: lib/src/history/code_history_controller.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (Malarg) [reply lib/src/history/code_history_controller.dart:30]: Don't you think that this approach makes ambiguily while reading the code. If we rename this field to `recordList` and methods like `saveRecord`, `removeRecords` and so on, code will be core clear to understand.

File: lib/src/history/code_history_controller.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply lib/src/history/code_history_controller.dart]: `+`

**Review by Malarg — APPROVED** [review]

File: lib/src/history/code_history_controller.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply lib/src/history/code_history_controller.dart]: `+`
